lua/ttt2/libraries/marker_vision.lua

```diff
@@ -0,0 +1,402 @@
+---
+-- A module that let's devs create bomb visions for any entity that is automatically synced
+-- between server and client and is tied to a player and their team
+-- @author Mineotopia
+-- @module markerVision
+
+if SERVER then
+	AddCSLuaFile()
+
+	util.AddNetworkString("ttt2_register_entity_removed")
+end
+
+VISIBLE_FOR_PLAYER = 0
+VISIBLE_FOR_TEAM = 1
+VISIBLE_FOR_ALL = 2
+
+VISIBLE_FOR_BITS = 3
+
+markerVision = {}
+
+markerVision.registry = {}
+
+---
+-- Registers an entity that should be rendered on screen with a wallhack and an optional
+-- UI element that works similar to targetID.
+-- @param Entity ent The entity that should be rendered
+-- @param Player owner The owner of the wallhack that takes their ownershipo with them on team change
+-- @param number visibleFor Visibility setting: `VISIBLE_FOR_PLAYER`, `VISIBLE_FOR_TEAM`, `VISIBLE_FOR_ALL`
+-- @param[opt] Color color The color of the wallhack, uses team color as fallback
+-- @param[opt] table receiverList A list of players that should receive the netmessage, overwrites the default
+-- @param[opt] any passThroughData any data that should be added to this radar vision, it is also synced to the client
+-- @note Call on server to add entity on server and all defined clients.
+-- @realm shared
+function markerVision.RegisterEntity(ent, owner, visibleFor, color, receiverList, passThroughData)
+	if SERVER then
+		local plysTeam = GetTeamFilter(owner:GetTeam(), true, true)
+
+		if #plysTeam == 0 then -- happens for TEAM_NONE for example
+			visibleFor = VISIBLE_FOR_PLAYER
+		end
+
+		local streamTable = {
+			ent = ent,
+			owner = owner,
+			visibleFor = visibleFor,
+			color = color,
+			data = passThroughData
+		}
+
+		local receiver
+
+		if visibleFor == VISIBLE_FOR_PLAYER then
+			receiver = owner
+		elseif visibleFor == VISIBLE_FOR_TEAM then
+			receiver = plysTeam
+		end
+
+		-- note: when VISIBLE_FOR_ALL receiver is nil, therefore SendStream uses net.Boradcast
+		net.SendStream("ttt2_register_entity_added", streamTable, receiverList or receiver)
+	end
+
+	if CLIENT then
+		if visibleFor == VISIBLE_FOR_ALL then
+			color = color or TEAMS[TEAM_INNOCENT].color
+		else
+			color = color or TEAMS[LocalPlayer():GetTeam()].color
+		end
```

Comment:
We do have translation strings for all 3 cases, so we should reflect these in color as well.

would the `TEAM_NONE.color` work for the `VISIBLE_FOR_PLAYER` case, wdyt?

Reply:
I think that would be confusing, wouldn't it? If something is visible only to you, then you'd expect it in your team color I think.

Also this would still clash with the unknown role etc

Reply:
no it's the other way around IMO. If i see something in my team's color i expect my whole team to see it.

Tbh the 'clash' with the unknown role is non-existent as the unknown doesn't have a team in gameplay terms and therefore does not care if a marker is only for them or for their "Team".

Reply:
I see your point now, thanks for raising the issue